Rerender send later options at specific times.

shameondev opened the pull request:
This PR introduces two changes. 

Firstly, it extracts the send later options to a separate template, allowing for the rendering of options for the send-later modal separately. This is necessary to ensure that the user can keep the modal open and still have access to actual sending options.

Secondly, it ensures that the send later options are re-rendered at a specific time.

Fixes: #25451.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details>
<summary>rerender at 15:55</summary>
<br>

![15:55](https://github.com/zulip/zulip/assets/53193850/f7b6857d-b57b-4e63-b826-4de4e9591ffb)

</details>

<details>
<summary>rerender at 08:55</summary>
<br>

![8:55](https://github.com/zulip/zulip/assets/53193850/4620a94a-d38c-47f4-a89b-68515cd219ae)

</details>

<details>
<summary>rerender at midnight</summary>
<br>

![midnight](https://github.com/zulip/zulip/assets/53193850/5df267e1-bee9-4435-8188-c50ebc96a82b)

</details>

karlstolley commented on web/src/scheduled_messages.js:
```diff
@@ -315,3 +317,29 @@ export function initialize(scheduled_messages_params) {
 }
+
+// exporting for testability
+export const _test_should_update_send_later_options = should_update_send_later_options;
```
You should able to test `should_update_send_later_options()` directly; there's no need I can see to export the function reference as a separate variable.

shameondev replied:
Sure!

shameondev commented on web/src/scheduled_messages.js:
```diff
@@ -315,3 +317,29 @@ export function initialize(scheduled_messages_params) {
+    }
+}
+
+setInterval(update_send_later_options, 1000 * 60);
```
> I wonder if it's worth setting the interval only when the modal opens

That will definitely be worth it! It will help eliminate the need for `is_send_later_modal_open`, at the very least. Thanks!

> you could have a const like SCHEDULING_MODAL_UPDATE_INTERVAL_IN_MILLISECONDS

Great, thanks!

shameondev commented on web/src/scheduled_messages.js:
```diff
@@ -315,3 +317,29 @@ export function initialize(scheduled_messages_params) {
+
+function update_send_later_options() {
+    const now = new Date();
+    const is_need_to_rerender = is_send_later_modal_open() && should_update_send_later_options(now);
```
Sure, thanks!

shameondev commented:
Hey @karlstolley, thanks a lot for your review! Your feedback was really helpful, and I've addressed all the points you raised. Would you mind taking a look again?

karlstolley commented on web/src/scheduled_messages.js:
```diff
@@ -16,6 +17,8 @@ import * as stream_data from "./stream_data";
 import * as timerender from "./timerender";
 
 export const MINIMUM_SCHEDULED_MESSAGE_DELAY_SECONDS = 5 * 60;
+export const SCHEDULING_MODAL_UPDATE_INTERVAL_IN_MILLISECONDS = 1000 * 60;
```
Since it's 60 seconds * 1000 milliseconds, I'd flip these two numbers (like the line above, with 5 minutes * 60 seconds).

karlstolley commented:
Just one little nit, but I think I've said all I have to say on this one :)

timabbott commented:
I made the following changes to further clean up the magic constants whose relationship between each other should be defined in the code:

``` diff
diff --git a/web/src/scheduled_messages.js b/web/src/scheduled_messages.js
index b13d5a52e3..0f7ff3e881 100644
--- a/web/src/scheduled_messages.js
+++ b/web/src/scheduled_messages.js
@@ -17,7 +17,7 @@ import * as stream_data from "./stream_data";
 import * as timerender from "./timerender";
 
 export const MINIMUM_SCHEDULED_MESSAGE_DELAY_SECONDS = 5 * 60;
-export const SCHEDULING_MODAL_UPDATE_INTERVAL_IN_MILLISECONDS = 1000 * 60;
+export const SCHEDULING_MODAL_UPDATE_INTERVAL_IN_MILLISECONDS = 60 * 1000;
 
 export let scheduled_messages_data = [];
 
@@ -330,8 +330,16 @@ export function initialize(scheduled_messages_params) {
 export function should_update_send_later_options(date) {
     const current_minute = date.getMinutes();
     const current_hour = date.getHours();
-    // We need to rerender if it is either 5 minutes before the hour or at midnight.
-    return current_minute === 55 || (current_hour === 0 && current_minute === 0);
+
+    if (current_hour === 0 && current_minute === 0) {
+        // We need to rerender the available options at midnight,
+        // since Monday could become in range.
+        return true;
+    }
+
+    // Rerender at MINIMUM_SCHEDULED_MESSAGE_DELAY_SECONDS before the
+    // hour, so we don't offer a 4:00PM send time at 3:59 PM.
+    return current_minute === 60 - MINIMUM_SCHEDULED_MESSAGE_DELAY_SECONDS / 60;
 }
 
 export function update_send_later_options() {
```

With those changes, merged, thanks @shameondev!

timabbott commented:
It seems possible we should rename `MINIMUM_SCHEDULED_MESSAGE_DELAY_SECONDS` to `MINIMUM_SCHEDULED_MESSAGE_DELAY_MINUTES`, but I'm not sure.

One thing I've been considering vaguely is whether we should arrange to have that value be 1 minute in the development environment to simplify manual testing.
